File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
answer = 1e9
def dfs(begin, target, words, cnt, visited):  #백트래킹
    global answer 
    if begin == target:
        answer = min(cnt, answer)
        return 
    for i in range(len(words)):
        if not visited[i]:
            if changeable(begin, words[i]):
                visited[i] = 1
                dfs(words[i], target, words, cnt + 1, visited)
                visited[i] = 0

#변경할 수 있는 단어인지 확인                 
def changeable(begin, target):
    diffCnt = 0 
    for i in range(len(begin)):
        if begin[i] != target[i]:
            diffCnt +=1
    
    if diffCnt == 1:
        return True
    else:
        return False
    
def solution(begin, target, words):
    visited = [0] * len(words)
    dfs(begin, target, words, 0, visited)
    if answer == 1e9 : return 0
    else : return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;
class Solution {
    static int answer;
    static boolean [] visited;
    static int n;
    static void DFS(String begin, String target, String[] words, int cnt){
        if (begin.equals(target)){
            answer = Math.min(answer, cnt);
            return;
        }
        
        for (int i = 0; i < n; i++) {
            if (!visited[i] && changeAble(begin, words[i])){
                visited[i] = true;
                DFS(words[i], target, words, cnt + 1);
                visited[i] = false;
            }
        }
    }
    static boolean changeAble(String begin, String target){
        char [] start = begin.toCharArray();
        char [] end = target.toCharArray();
        int cnt = 0;
        for (int i = 0; i < start.length; i++){
            if (start[i] != end[i]) cnt ++;
        }
        if (cnt == 1)
            return true;
        else 
            return false;
    }
    public int solution(String begin, String target, String[] words) {
        answer = 1000000000;
        n = words.length;
        visited = new boolean[n];
        DFS(begin, target, words, 0);
        
        if (answer == 1000000000) answer = 0;
        return answer;
    }
}